File: Components/cross_point_uniform.m
function [output1,output2] = cross_point_uniform(varargin)
% Uniform crossover.

%------------------------------Copyright-----------------------------------
% Copyright (C) <2023>  <Qi Zhao>

% AutoOptLib is a free software. You can use, redistribute, and/or modify
% it under the terms of the GNU General Public License as published by the 
% Free Software Foundation, either version 3 of the License, or any later 
% version. 

% Please reference the paper below if using AutoOptLib in your publication:
% @article{zhao2023autooptlib,
%  title={AutoOptLib: A Library of Automatically Designing Metaheuristic 
%         Optimization Algorithms in Matlab},
%  author={Zhao, Qi and Yan, Bai and Hu, Taiwei and Chen, Xianglong and 
%          Yang, Jian and Shi, Yuhui},
%  journal={arXiv preprint 	arXiv:2303.06536},
%  year={2023}
% }
%--------------------------------------------------------------------------


mode = varargin{end};
switch mode
    case 'execute'
        Parent = varargin{1};
        Para   = varargin{3};
        Aux    = varargin{4};

        Prob = Para;
        Parent = Parent.decs;
        [N,D] = size(Parent);
        Parent1 = Parent(1:ceil(N/2),:);
        Parent2 = Parent(floor(N/2)+1:end,:);
        Nhalf = size(Parent1,1);

        ind = rand(Nhalf,D) < Prob;
        Offspring1 = Parent1;
        Offspring2 = Parent2;
        Offspring1(ind) = Parent2(ind);
        Offspring2(ind) = Parent1(ind);
        Offspring = [Offspring1;Offspring2];
        output1   = Offspring(1:N,:);
        output2   = Aux;

    case 'parameter'
        output1 = [0,0.5]; % crossover probability

    case 'behavior'
        output1 = {'LS','small';'GS','large'}; % small probabilities perform local search
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: AutoOptLib for Python/components/cross_point_uniform.py
"""Python translation of cross_point_uniform."""

from __future__ import annotations

import numpy as np


def cross_point_uniform(*args):
    mode = args[-1]
    if mode == "execute":
        parent = args[0]
        para = args[2] if len(args) > 2 else None
        aux = args[3] if len(args) > 3 else None  # noqa: F841
        prob = float(np.asarray(para).reshape(-1)[0]) if para is not None else 0.5
        decs = getattr(parent, "decs", None)
        if callable(decs):
            parent = decs()
        else:
            parent = decs if decs is not None else parent
        parent = np.asarray(parent)
        n, d = parent.shape
        p1 = parent[: (n + 1) // 2]
        p2 = parent[n // 2 :]
        nh = p1.shape[0]
        mask = np.random.rand(nh, d) < prob
        off1 = p1.copy()
        off2 = p2.copy()
        off1[mask] = p2[mask]
        off2[mask] = p1[mask]
        offspring = np.vstack([off1, off2])
        return offspring[:n, :], args[3] if len(args) > 3 else None
    if mode == "parameter":
        return [0, 0.5], None
    if mode == "behavior":
        return [["LS", "small"], ["GS", "large"]], None
    raise ValueError(f"Unsupported mode: {mode}")
